### PR TITLE
Fail wait_for_ready=false RPCs when channel is in TRANSIENT_FAILURE.

### PR DESCRIPTION
**Pull request by markdroth**
Let's see how many tests break...

### CODE REVIEW

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                                             FILE SIZE
 ++++++++++++++ GROWING                                               ++++++++++++++
  +0.7%    +256 src/core/ext/filters/client_channel/client_channel.cc    +256  +0.7%
       +17%    +246 cc_start_transport_stream_op_batch                       +246   +17%
      +2.7%     +10 [Unmapped]                                                +10  +2.7%
  +0.0%     +96 [None]                                                   +184  +0.0%

  +0.0%    +352 TOTAL                                                    +440  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                                             FILE SIZE
 ++++++++++++++ GROWING                                               ++++++++++++++
  +0.7%    +256 src/core/ext/filters/client_channel/client_channel.cc    +256  +0.7%
       +17%    +246 cc_start_transport_stream_op_batch                       +246   +17%
      +2.7%     +10 [Unmapped]                                                +10  +2.7%
  +0.0%     +96 [None]                                                   +184  +0.0%

  +0.0%    +352 TOTAL                                                    +440  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,966,675      Total (>)      1,966,345

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,909,773      Total (>)     10,909,771

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,966,675      Total (>)      1,966,345

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,909,775      Total (>)     10,909,772

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                                                                            FILE SIZE
 ++++++++++++++ GROWING                                                                              ++++++++++++++
  +0.0%    +104 [None]                                                                                  +472  +0.0%
  +0.5%    +192 src/core/ext/filters/client_channel/client_channel.cc                                   +192  +0.5%
      [NEW]    +389 add_closure_for_subchannel_batch(grpc_call_element*, grpc_transport_stream_op_batch*    +389  [NEW]
      [NEW]    +203 fail_call_if_in_transient_failure                                                       +203  [NEW]
      [NEW]    +192 add_retriable_send_message_op(grpc_call_element*, subchannel_call_retry_state*, subc    +192  [NEW]
      +1.1%      +4 [Unmapped]                                                                                +4  +1.1%
      +0.1%      +1 start_pick_locked                                                                         +1  +0.1%

  +0.0%    +296 TOTAL                                                                                   +664  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,997,560      Total (>)      1,997,294

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
11,021,964      Total (>)     11,021,759

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
Corrupt JSON data (indicates timeout or crash): 
    bm_call_create.BM_IsolatedFilter_ClientChannelFilter_NoOp_.counters.new: 10
    bm_call_create.BM_IsolatedFilter_ClientChannelFilter_NoOp_.counters.old: 10


[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                                                                            FILE SIZE
 ++++++++++++++ GROWING                                                                              ++++++++++++++
  +0.0%    +104 [None]                                                                                  +472  +0.0%
  +0.5%    +192 src/core/ext/filters/client_channel/client_channel.cc                                   +192  +0.5%
      [NEW]    +389 add_closure_for_subchannel_batch(grpc_call_element*, grpc_transport_stream_op_batch*    +389  [NEW]
      [NEW]    +203 fail_call_if_in_transient_failure                                                       +203  [NEW]
      [NEW]    +192 add_retriable_send_message_op(grpc_call_element*, subchannel_call_retry_state*, subc    +192  [NEW]
      +1.1%      +4 [Unmapped]                                                                                +4  +1.1%
      +0.1%      +1 start_pick_locked                                                                         +1  +0.1%

  +0.0%    +296 TOTAL                                                                                   +664  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,997,560      Total (>)      1,997,294

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
11,021,961      Total (>)     11,021,758

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
Corrupt JSON data (indicates timeout or crash): 
    bm_call_create.BM_IsolatedFilter_ClientChannelFilter_NoOp_.counters.new: 10
    bm_call_create.BM_IsolatedFilter_ClientChannelFilter_NoOp_.counters.old: 10


[microbenchmarks] No significant performance differences
```

**Comment by markdroth** [comment]
@AspirinSJL, please take another look, since I changed the code to defer the check until after the service config is applied.

File: src/core/ext/filters/client_channel/client_channel.cc [hunk 1]
```diff
@@ -3127,6 +3151,16 @@ static void start_pick_locked(void* arg, grpc_error* ignored) {
     // We do not yet have an LB policy, so wait for a resolver result.
     if (GPR_UNLIKELY(!chand->started_resolving)) {
       start_resolving_locked(chand);
+    } else {
+      // Normally, we want to do this check in
+      // process_service_config_and_start_lb_pick_locked(), so that we
+      // can honor the wait_for_ready setting in the service config.
+      // However, if the channel is in TRANSIENT_FAILURE at this point, that
+      // means that the resolver has returned a failure, so we're not going
```
Comment (AspirinSJL) [inline src/core/ext/filters/client_channel/client_channel.cc]: Maybe the resolver is just slow instead of having returned a failure?

Comment (markdroth) [reply src/core/ext/filters/client_channel/client_channel.cc]: The channel won't be in TRANSIENT_FAILURE in that case; it will still be in CONNECTING.

**Comment by AspirinSJL** [comment]
LGTM

**Comment by markdroth** [comment]
Known issues: #16985

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                                                                            FILE SIZE
 ++++++++++++++ GROWING                                                                              ++++++++++++++
  +0.0%    +104 [None]                                                                                  +464  +0.0%
  +0.5%    +192 src/core/ext/filters/client_channel/client_channel.cc                                   +192  +0.5%
      [NEW]    +389 add_closure_for_subchannel_batch(grpc_call_element*, grpc_transport_stream_op_batch*    +389  [NEW]
      [NEW]    +203 fail_call_if_in_transient_failure                                                       +203  [NEW]
      [NEW]    +192 add_retriable_send_message_op(grpc_call_element*, subchannel_call_retry_state*, subc    +192  [NEW]
      +1.1%      +4 [Unmapped]                                                                                +4  +1.1%
      +0.1%      +1 start_pick_locked                                                                         +1  +0.1%

  +0.0%    +296 TOTAL                                                                                   +656  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,997,540      Total (>)      1,997,274

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
11,021,901      Total (>)     11,021,697

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
Corrupt JSON data (indicates timeout or crash): 
    bm_call_create.BM_IsolatedFilter_ClientChannelFilter_NoOp_.counters.new: 10
    bm_call_create.BM_IsolatedFilter_ClientChannelFilter_NoOp_.counters.old: 10


[microbenchmarks] No significant performance differences
```

**Pull request merged**